scripts/calculateRevenue/protocols/utils/events.ts

```diff
@@ -42,13 +42,94 @@ export async function _getNearestBlock(
     )
   }
   const blockTimestampData = (await response.json()) as BlockTimestampData
-  return blockTimestampData.height
+  return blockTimestampData
 }
 
-export const getNearestBlock = memoize(_getNearestBlock, {
+/**
+ * Intentionally not exported. We should use `getBlockRange` instead.
+ */
+const getNearestBlock = memoize(_getNearestBlock, {
   hash: (...params: Parameters<typeof _getNearestBlock>) => params.join(','),
 })
 
+export const _getNearestBlockForTesting = getNearestBlock
+
+async function getFirstBlockAtOrAfterTimestamp(
+  networkId: NetworkId,
+  targetDate: Date,
+): Promise<number> {
+  const targetTimestampSec = Math.floor(targetDate.getTime() / 1000)
+  const block = await getNearestBlock(networkId, targetDate)
+
+  if (block.timestamp >= targetTimestampSec) {
+    // The nearest block is at or after the target. It must be the first.
+    return block.height
+  } else {
+    // The nearest block is before the target. The next block must be the first one at or after.
+    // Note: Assumes block numbers increment by 1 and block N+1 always has timestamp >= block N.
+    return block.height + 1
+  }
+}
+
+/**
+ * Calculates the start and end block numbers for a given time range.
+ * The time range is defined as [startTimestamp, endTimestamp),
+ * meaning it's inclusive of the startTimestamp and exclusive of the endTimestamp.
+ *
+ * @param networkId The ID of the network.
+ * @param startTimestamp The inclusive start date of the time range.
+ * @param endTimestamp The exclusive end date of the time range.
+ * @returns A promise that resolves to an object containing:
+ *    `startBlock`: The first block whose timestamp is >= startTimestamp (inclusive).
+ *    `endBlock`: The first block whose timestamp is >= endTimestamp. This block itself
+ *                is *exclusive* from the desired range. When used in loops like
+ *                `for (let i = startBlock; i < endBlock; i++)`, or as an exclusive
+ *                upper bound in queries, it correctly defines the desired time window.
+ * @throws Will throw an error if startTimestamp is not before endTimestamp, or if a valid
+ *         block range cannot be determined (e.g., startBlock ends up >= endBlock).
+ */
+export async function getBlockRange({
+  networkId,
+  startTimestamp,
+  endTimestamp,
```

Comment:
I'm considering renaming this to `endTimestampExclusive` for maximum clarity too. In a follow up PR.
What do you think?

Reply:
sounds consistent! 👍🏻